src/python/pants/backend/java/target_types.py

```diff
@@ -24,6 +41,23 @@ class JunitTests(Target):
     help = "Java tests, run with Junit."
 
 
+class GenerateJunitTestsFromJunitTests(GenerateTargetsRequest):
+    generate_from = JunitTests
+
+
+@rule
+async def generate_junit_tests_from_junit_tests(
+    request: GenerateJunitTestsFromJunitTests, union_membership: UnionMembership
+) -> GeneratedTargets:
+    paths = await Get(SourcesPaths, SourcesPathsRequest(request.generator[JavaTestsSources]))
+    return generate_file_level_targets(JunitTests, request.generator, paths.files, union_membership)
```

Comment:
In terms of addressing this boilerplate: is it worth having two triggers: one union, and one bool or sentinel union? The bool or sentinel would request a built-in `Type1->Type1` generation, and the union would allow a `Type1->Type2` conversion and arbitrary logic, but would require a `@rule`?

Reply:
I don't think the boilerplate is worth worrying about yet. The followup change to move where sibling-file dependencies are injected makes this much more clear. Callers start to diverge with that change in a meaningful way.